pkg/river/token/builder/builder_test.go

```diff
@@ -409,3 +409,266 @@ func TestBuilder_MapBlocks(t *testing.T) {
+
+func TestBuilder_AppendInner_DiffersFromDefault(t *testing.T) {
+	type Structure struct {
+		Field          string                 `river:"field,attr"`
+		Block          InnerBlockDefaulter    `river:"block,block,optional"`
+		BlocksSlice    []InnerBlockDefaulter  `river:"block_slice,block,optional"`
+		BlockPtr       *InnerBlockDefaulter   `river:"block_ptr,block,optional"`
+		BlocksPtrSlice []*InnerBlockDefaulter `river:"block_ptr_slice,block,optional"`
+	}
+
+	f := builder.NewFile()
+	f.Body().AppendFrom(Structure{
+		Field: "some_value",
+
+		Block: InnerBlockDefaulter{Number: 1},
+		BlocksSlice: []InnerBlockDefaulter{
+			{Number: 1}, {Number: 2},
+		},
+
+		BlockPtr: &InnerBlockDefaulter{Number: 1},
+		BlocksPtrSlice: []*InnerBlockDefaulter{
+			{Number: 1}, {Number: 2},
+		},
+	})
+
+	expect := format(t, `
+		field = "some_value"
+		
+		block {
+			number = 1
+		}
+		
+		block_slice {
+			number = 1
+		}
+		
+		block_slice {
+			number = 2
+		}
+		
+		block_ptr {
+			number = 1
+		}
+		
+		block_ptr_slice {
+			number = 1
+		}
+		
+		block_ptr_slice {
+			number = 2
+		}
+	`)
+
+	require.Equal(t, expect, string(f.Bytes()))
+}
+
+func TestBuilder_AppendInner_MatchesDefaults(t *testing.T) {
+	type Structure struct {
+		Field          string                 `river:"field,attr"`
+		Block          InnerBlockDefaulter    `river:"block,block,optional"`
+		BlocksSlice    []InnerBlockDefaulter  `river:"block_slice,block,optional"`
+		BlockPtr       *InnerBlockDefaulter   `river:"block_ptr,block,optional"`
+		BlocksPtrSlice []*InnerBlockDefaulter `river:"block_ptr_slice,block,optional"`
+	}
+
+	f := builder.NewFile()
+	f.Body().AppendFrom(Structure{
+		Field: "some_value",
+
+		Block: InnerBlockDefaulter{Number: 123},
+		BlocksSlice: []InnerBlockDefaulter{
+			{Number: 123}, {Number: 123},
+		},
+
+		BlockPtr: &InnerBlockDefaulter{Number: 123},
+		BlocksPtrSlice: []*InnerBlockDefaulter{
+			{Number: 123}, {Number: 123},
+		},
+	})
+
+	expect := format(t, `
+		field = "some_value"
+		
+		block { }
+		
+		block_slice { }
+		
+		block_slice { }
+		
+		block_ptr { }
+		
+		block_ptr_slice { }
+		
+		block_ptr_slice { }
+
+	`)
+
+	require.Equal(t, expect, string(f.Bytes()))
+}
+
+func TestBuilder_AppendInner_EmptyValues(t *testing.T) {
+	type Structure struct {
+		Field          string                 `river:"field,attr"`
+		Block          InnerBlockDefaulter    `river:"block,block,optional"`
+		BlocksSlice    []InnerBlockDefaulter  `river:"block_slice,block,optional"`
+		BlockPtr       *InnerBlockDefaulter   `river:"block_ptr,block,optional"`
+		BlocksPtrSlice []*InnerBlockDefaulter `river:"block_ptr_slice,block,optional"`
+	}
+
+	f := builder.NewFile()
+	f.Body().AppendFrom(Structure{
+		Field: "some_value",
+
+		Block:       InnerBlockDefaulter{},
+		BlocksSlice: []InnerBlockDefaulter{},
+
+		BlockPtr:       &InnerBlockDefaulter{},
+		BlocksPtrSlice: []*InnerBlockDefaulter{},
+	})
+
+	expect := format(t, `
+		field = "some_value"
+
+		block_ptr { }
```

Comment:
Why are all the values skipped? they all should have `number = 0`. Looks like a bug?

Reply:
Debugging this for the field named `block`:
![image](https://github.com/grafana/agent/assets/17101802/9c629adb-f261-416b-b8be-e2d87c808dc0)
We can see that it's determined to match the default:
![image](https://github.com/grafana/agent/assets/17101802/fcf3e108-2ce4-4f2b-b8a4-c1edcb460b39)

But that's the default of the `Structure`, not the `InnerBlockDefaulter`. The correct default for the `InnerBlockDefaulter` has `123` in it.

I think that's the issue. For a given field in an inner block we have two possible default values:
* One coming from the outer block's `SetToDefault`
* Another coming from the fields' own `SetToDefault`

They can give conflicting defaults. We also need to be careful with how we treat `nil` returned in defaults.

Reply:
I've drafted a fix here, but this may not be the best way to tackle it: https://github.com/grafana/agent/pull/4435/files